drivers/wifi/esp8266/wifi_esp8266.c

```diff
@@ -0,0 +1,1570 @@
+		CMD_HANDLER("SEND OK", sendok),
+		CMD_HANDLER("link is not valid", atcmdecho_nosock),
+		CMD_HANDLER("busy p...", atcmdecho_nosock),
+		CMD_HANDLER("busy s...", atcmdecho_nosock),
+		CMD_HANDLER("ready", esp8266_ready),
+		CMD_HANDLER("AT+CIPAPMAC_CUR?", atcmdecho_nosock),
+		CMD_HANDLER("+CIPAPMAC_CUR:", mac_addr_resp),
+		CMD_HANDLER("AT+CIPSTA_CUR?", atcmdecho_nosock),
+		CMD_HANDLER("+CIPSTA_CUR:", ip_addr_resp),
+		CMD_HANDLER("AT+CWLAP", atcmdecho_nosock),
+		CMD_HANDLER("+CWLAP:", wifi_scan_resp),
+		CMD_HANDLER("+CWLAP:", wifi_scan_resp),
+		CMD_HANDLER("AT+CIPMUX", atcmdecho_nosock),
+		CMD_HANDLER("AT+CWQAP", atcmdecho_nosock),
+		CMD_HANDLER("AT+CIPSEND=", atcmdecho_nosock),
+		CMD_HANDLER_NO_SKIP("0,CONNECT", conn_open),
+		CMD_HANDLER_NO_SKIP("1,CONNECT", conn_open),
+		CMD_HANDLER_NO_SKIP("2,CONNECT", conn_open),
+		CMD_HANDLER_NO_SKIP("3,CONNECT", conn_open),
+		CMD_HANDLER_NO_SKIP("4,CONNECT", conn_open),
+		CMD_HANDLER_NO_SKIP("0,CLOSED", conn_closed),
+		CMD_HANDLER_NO_SKIP("1,CLOSED", conn_closed),
+		CMD_HANDLER_NO_SKIP("2,CLOSED", conn_closed),
+		CMD_HANDLER_NO_SKIP("3,CLOSED", conn_closed),
+		CMD_HANDLER_NO_SKIP("4,CLOSED", conn_closed),
+		CMD_HANDLER("Recv ", sent_bytes),
+	};
+
+	while (true) {
+		k_sem_take(&foo_data.mdm_ctx.rx_sem, K_FOREVER);
+
+		esp8266_read_rx(&rx_buf);
+
+		while (rx_buf) {
+
+			if (foo_data.data_len) {
+				if (net_buf_frags_len(rx_buf) >=
+					foo_data.data_len) {
+					esp8266_read_data(&rx_buf);
+
+					if (!rx_buf) {
+						break;
+					}
+					continue;
+				} else {
+						break;
+				}
+			}
+
+			net_buf_skipcrlf(&rx_buf);
+			if (!rx_buf) {
+				break;
+			}
+
+
+			/* check for incoming IPD data */
+			if (match(rx_buf, "+IPD") >= 0) {
+				if (esp8266_process_setup_read(&rx_buf) < 0) {
+					break;
+				}
+
+				if (!rx_buf) {
+					break;
+				}
+
+				continue;
+			}
+
+			if (match(rx_buf, ">") >= 0) {
+				rx_buf = net_buf_skip(rx_buf, 1);
+				k_sem_give(&foo_data.response_sem);
+			}
+
+			frag = NULL;
+			len = net_buf_findcrlf(rx_buf, &frag, &offset);
+			if (!frag) {
+				break;
+			}
+
+			/* look for matching data handlers */
+			i = -1;
+			for (i = 0; i < ARRAY_SIZE(handlers); i++) {
+				if (match(rx_buf, handlers[i].cmd) >= 0) {
+
+					/* found a matching handler */
+					LOG_INF("MATCH %s (len:%u)\n",
+						    handlers[i].cmd, len);
+
+					/* skip cmd_len */
+					if (handlers[i].skip) {
+						rx_buf = net_buf_skip(rx_buf,
+								handlers[i].cmd_len);
+					}
+
+					/* locate next cr/lf */
+					frag = NULL;
+					len = net_buf_findcrlf(rx_buf,
+							       &frag, &offset);
+
+					/* call handler */
+					if (handlers[i].func) {
+						handlers[i].func(&rx_buf, len);
+					}
+
+					frag = NULL;
+					/* make sure buf still has data */
+					if (!rx_buf) {
+						break;
+					}
+
+					/* locate next cr/lf */
+					len = net_buf_findcrlf(rx_buf,
+							       &frag, &offset);
+					if (!frag) {
+						break;
+					}
+
+					break;
+				}
+			}
```

Comment:
way too many while/if/for (up to 5, where our code style - same as linux - tries to limit these to 3) here, try to factorize the code for better clarity.

Reply:
well, it's basically a parser looking for conditions and there are quirks to the way the ESP8266 handles data send and receive (the raw packet data).  I'll have to see how pulling out some of this and generalizing it for this driver and the other driver that uses a lot of the same types of things works out.